### PR TITLE
WIP: internal/FormatWriter simplify two regexen; fix pipe char bug

### PR DESCRIPTION
**Pull request by LeeTibbert**
* This PR was motivated by recent discussions related to Scalafmt
    PRs #1916, #1917, and those mentioned in those PRs.

  * The primary purpose of this PR is to provide a concrete example
    of how the regular expressions for `leadingAsteriskSpace` and
    `compileStripMarginPattern` be simplified.

    Discussions about readability are always subjective. I believe
    that removing the positive and negative lookaheads makes it
    more likely that the next maintainer will be able to read and
    comprehend the code.

    Discussions about runtime performance improvements without
    demonstrable data are worth than useless.  lookaheads are
    known to have a high runtime cost. It is believable, but not
    asserted by this PR for lack of evidence, that the code of this
    PR will have better performance characteristics.

  * Whilst simplifying the two regex expressions, I noticed what I
    believe to be a bug in `compileStripMarginPattern`. Scala allows
    any character other than NULL to be used as a pipe character.
    Java requires that Dollar ($) and backslash (\) characters
    in the replacement string be quoted (e.g. in a string: "\\$").
    The code prior to this PR did not do this.
  
  * The handling of `compileStripMarginPattern` is a bit tricky. Every/any regex expression should
     be sealed with the Curse of the Mummy, but that one is particularly deserving.

     The active pipe character is not in scope where the replaceAll() is done. 
     Using a capture group allows the pipe character from where the pattern is compiled to be
     used in replaceAll(),
     
     Agreed, a code smell, but, as a guest, I wanted/needed to minimize changes to the 
     existing code. One capture group is not all that expensive. The "[^*]" character class
     in `leadingAsteriskSpace` probably takes more memory.

  * The code of this PR passes all current scalafmt tests:
    `sbt> tests/test`
 ```
[info] ScalaTest
[info] Run completed in 1 minute, 23 seconds.
[info] Total number of tests run: 2406
[info] Suites: completed 18, aborted 0
[info] Tests: succeeded 2406, failed 0, canceled 0, ignored 8, pending 0
[info] All tests passed.
[info] Passed: Total 2406, Failed 0, Errors 0, Passed 2406, Ignored 8
[success] Total time: 90 s (01:30), completed Apr 28, 2020, 2:41:12 PM
 ```

  * Three cases were added to  `./scalafmt-tests/src/test/resources/test/StripMargin.stat` to 
     ensure that potentially troublesome stripMargin() pipe characters, such as Dollar Sign,
     Backslash, and Right Parenthesis work as intended.
 
     Test cases for Unicode characters in leadingAsteriskSpace for trailing whitespace
     remain as an exercise for the reader. The lack of such tests is a blemish on my
     engineering.

 * I did not test it but the `\n` approach should work on the/most Windows operating system.
   It is what has been in the code for years.

 * The regular expressions in this PR should work with Scala Native and make an eventual
    port easier.

Documentation:

  * None required. The changes in this PR should cause no user visible
    change. I do not know if the practice of scalafmt development is to create
    a change log entry for user invisible changes or if the git history is enough.

Testing:

  * Built from a clean slate and manually tested ("test/test")
    using sbt 1.3.10 & Java 11 on X86_64 only. All tests pass.

### CODE REVIEW

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 1]
```diff
@@ -735,7 +737,7 @@ object FormatWriter {
   }
 
   private val leadingAsteriskSpace =
-    Pattern.compile("(?<=\\n)\\h*(?=[*][^*])")
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: Can you explain why allowing match on multiple asterisks is equivalent? I tried yesterday, there was a test which failed but looks like for you it didn't.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 2]
```diff
@@ -758,7 +762,8 @@ object FormatWriter {
     if (pipe == '|') leadingPipeSpace else compileStripMarginPattern(pipe)
   @inline
   private def compileStripMarginPattern(pipe: Char) =
-    Pattern.compile(s"(?<=\\n)\\h*(?=[$pipe])")
+    // capture group is used to pass pipe character to distant replaceAll.
+    Pattern.compile(s"\n\\h*(\\${pipe})", Pattern.MULTILINE)
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:766]: What if pipe contains a closing parenthesis? I admit that my solution wouldn't work for pipe containing a closing bracket but I'm not sure why this approach is fully insulated.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 3]
```diff
@@ -735,7 +737,7 @@ object FormatWriter {
   }
 
   private val leadingAsteriskSpace =
-    Pattern.compile("(?<=\\n)\\h*(?=[*][^*])")
+    Pattern.compile("\n\\h*\\*([^*])", Pattern.MULTILINE)
```
[the remark this answers is not in the document]
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:740]: You are correct, no anchors are in use. I can try without the MULTILINE. Thank you for second set of eyes.

I had convinced myself that it was necessary but am probably all twisted up.
An experiment will soon verify.   Simpler if it can be removed.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 1 above)
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: I took the requirement that " *X" gets replaced and " **" (double asterisk) does not as
"Do not even _think_ about breaking this" lore.  The match in the capture group
should match anything but what would be a second asterisk [^*].

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 2 above)
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:766]: Challenge-response helps build better code.  You make a good (friendly argument) for
some explicit test cases. 

By my reading & intent, a closing parenthesis of any ilk: round, curly, square.
It should be escaped by the backslash (which, for future readers,
needs to be doubled because it is in an interpolated string). 

Once the pipe character is in the capture group, it should able to be used by
replaceAll with no problems.  Based on no data, that is my story..

Does it parse that way for you?

It is late in my day today and I have promises to keep but I'll take a run at stripMargin tests
which check out the usual suspects.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 1 above)
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: ah, was looking at the code on my phone, didn't see the `[^*]`. never mind.

`\n` vs `\\n`, any difference?

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: > `\n` vs `\\n`, any difference?

thanks for adding tests. this question above is the last one.

Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: re: `\n vs \\n, any difference?`   

Pardon me for being concrete, no disrespect intended.  The concrete answer
is more of my inability to organize my thoughts above that. I have spent 
hours, days, or, more probably weeks, being hosed by exactly this issue.
The recurrence interval between episodes of my being hosed is just slightly
longer than my memory of the solution.   Mea culpa!

It is my understanding that for 'normal' strings  `\n` is always used and gets translated
to one standard Java two-octet Character.

In s-interpolated strings (and possibly/probably other interpolations) things get
a bit more confusing, at least for me.  I believe that a `\n` gets translated as for
a simple string. In `\\n` the first backslash quotes the second backslash, which
then becomes `\n`  which Pattern.compile interprets as though the original
string had been `\n`.

So, both work in interpolated strings, unless you have evidence or experience to
the contrary.  I tend to prefer the single `\n`  because it has fewer characters and transformations to parse.

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: > re: `\n vs \\n, any difference?`
> 
> So, both work in interpolated strings, unless you have evidence or experience to
> the contrary. I tend to prefer the single `\n` because it has fewer characters and transformations to parse.

like `\\s`, `\\h` or other special escape sequences, i was curious whether `\\n` is interpreted by `Pattern` as anything different than simple `0xA`. clearly, `\n` is directly substituted as 0xA so pattern only sees that.

Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: Re: tests 

I found StripMargin.stat and added tests for only the `align` aspect. I did not address
the `no align` case, which is a weasel, driven by limited time.

I had expected to find tests for leading and trailing spaces. I could not find anything relevant
after a number of trials: using find for filenames and/or grepping for "whitespace" and
combinations.

I did find a Unicode.stat and could slip in a case for leading and trailing horizontal
whitespace containing Unicode there.  It would seem more integrated to have
it in a leadingWhitespace.spec if such a thing exists.  Would, IMHO, make it easier
for future visiting devos, such as myself, to find.

Off the top of your head, do you have any clues or suggestions.

Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: > Off the top of your head, do you have any clues or suggestions.

I added one tab in the `default/Comment` test, see at the bottom:

https://github.com/scalameta/scalafmt/pull/1911/files

you can create a dedicated whitespace test but i considered it part of handling comments and strings.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 2 above)
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala]: Fixed by adding three unit tests to StripMargin.stat for alternate pipe characters, including
Right or closing round parenthesis.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 1 above)
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:738]: @kitbellew The info you gave helped me to get to the next round.  Thank you.

All tests, new & previous,  pass on my system.  I will check CI later.

The tests are still sensitive to some future devo removing the now invisible whitespace
especially the tabs.

I think the three regex expressions are as simple & fast as the requirements and reasonable
end use allows.  I left a comment in FormatWriter.scala explaining why. Figured that might
endure long after the present conversation is forgotten.

I suspect that there are a couple of bugs remaining in scalafmt relating to tabs. I will see in
my mainline daily work  if I can trick them from coming out of the dim shadows.

Unit tests prove their worth, much as they are a pain to write.  The hard time I had
getting unit tests to accept Ogham marks and unicode \uMumble convinced me
that a simple "space or tab" test is all that is needed.  There is nothing wrong with
"\h", just that it will spend a lot of time testing against characters which can not be 
in the input.   Sorry if I lead you down a loosing path.

If you like the current work, I can update & buff the commit message and change the
title.

File: scalafmt-tests/src/test/resources/default/Comment.stat [hunk 4]
```diff
@@ -125,7 +125,7 @@ object a {
   /*
 
    /**
-     		* comment
```
Comment (kitbellew) [inline scalafmt-tests/src/test/resources/default/Comment.stat:128]: did you remove this because you create a separate test case for that?

File: scalafmt-tests/src/test/resources/spaces/TrailingWhiteSpace.stat [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+object B {
+  val a = 1	 	
+  val b = 2	 	
+}
+>>>
+object B {
+  val a = 1
+  val b = 2
+}
+
```
Comment (kitbellew) [inline scalafmt-tests/src/test/resources/spaces/TrailingWhiteSpace.stat:23]: would you mind remove the 3 blank lines at the end?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 6]
```diff
@@ -731,12 +731,30 @@ object FormatWriter {
...
+    * gives the recognized scala whitespace characters.
+    * "Whitespace characters. \U0020 | \U0009 | \U000D | \U000A"
+    * (original text has lowercase u. U substituted here to allow scalafmt
+    *  to format this file.)
+    *
+    * Of the characters which the parser will pass, only space (\u0020)
+    * and tab (\u0009) are horizontal whitespace.
+    * Save execution time by not checking what the character could not be.
+    *
+    * */
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:747]: should this be `*/` or `**/` instead of `* */`?

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 7]
```diff
@@ -731,12 +731,30 @@ object FormatWriter {
...
   }
 
-  private val trailingSpace = Pattern.compile("\\h+$", Pattern.MULTILINE)
+  /**
+    * No Unicode is necessary in the three (3) Pattern.compile() below.
+    * scalafmt parser will have already have objected.
+    * https://github.com/scala/scala/blob/2.11.x/spec/01-lexical-syntax.md
+    * gives the recognized scala whitespace characters.
+    * "Whitespace characters. \U0020 | \U0009 | \U000D | \U000A"
+    * (original text has lowercase u. U substituted here to allow scalafmt
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:740]: is this comment relevant to the code? :)

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala [hunk 8]
```diff
@@ -731,12 +731,30 @@ object FormatWriter {
     }
   }
 
-  private val trailingSpace = Pattern.compile("\\h+$", Pattern.MULTILINE)
+  /**
+    * No Unicode is necessary in the three (3) Pattern.compile() below.
+    * scalafmt parser will have already have objected.
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:736]: are you saying that nothing other than space and tab are allowed by scalac or scalameta parser that precedes formatting?

**Comment by kitbellew** [comment]
@LeeTibbert will you be able to update this soon? we are planning to release v2.5.0 this week.

@poslegm can you please review? this is a follow-up to #1911, #1916 and #1917.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 8 above)
Comment (poslegm) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:736]: It **shouldn't** be allowed according to specs. 

A cursory examination of the scalameta sources led to [this](https://github.com/scalameta/scalameta/blob/0fcba272a9f8d91b22fb8a180a805acf443a30be/scalameta/tokenizers/shared/src/main/scala/scala/meta/internal/tokenizers/Chars.scala#L99). So, it matches.

**Comment by poslegm** [comment]
@kitbellew I am ok with it. Also I think no one will be hurt if this goes into the 2.5.1 instead of 2.5.0 :) So no need to rush

File: scalafmt-tests/src/test/resources/default/Comment.stat (hunk 4 above)
Comment (LeeTibbert) [reply scalafmt-tests/src/test/resources/default/Comment.stat:128]: Exactly.   The case seemed to be multiplexed in a non-obvious (to me way).  
I de-multiplexed the concerns.

Creating a second test with a title which described its primary (and sole) concerned
seemed to be a gift to future people trying to suss this stuff.
The added complexity of a second test seemed worth the benefit.

File: scalafmt-tests/src/test/resources/spaces/TrailingWhiteSpace.stat (hunk 5 above)
Comment (LeeTibbert) [reply scalafmt-tests/src/test/resources/spaces/TrailingWhiteSpace.stat:23]: Done, thank you.  So you have something against trailing whitespace ;-) 
Guess I had expected scalafmt to remove those for me.  Must not be set in
the project .scalafmt.conf or I bungled it.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 6 above)
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:747]: Being a visitor and it not being my usual coding style, for consistency I did a 
"monkey see, monkey do" of several other comments in the same file.

Yes, the trailing "* */" is not a variant I have seen in recent memory,

That comment is re-worked in the commit I am about to submit, using my
usual "//" style.  If the concept of the comment is good, I can adapt the 
comment style to anything you suggest.  Small dragons.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 7 above)
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:740]: I had thought so, but that section is superseded now.  After I commit, let me know if
I can improve the new text.   I try to get enough across so that I can remember six months
down the road and so that a colleague can read it without complaining about "War & Peace".

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala (hunk 8 above)
Comment (LeeTibbert) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala:736]: After another long day of creating test cases and trying to be perverse, it turns out
that trailing whitespace with Unicode whitespace _is_ allowed in comments. As cited,
it is rejected in non-comment code.

I am about to do a commit which handles, tests, & succinctly (I hope) documents.

The discussion here helps.

**Comment by LeeTibbert** [comment]
If it does not cause unnecessary work at your end, I would be happier getting on the
`2.5.1 ` train. I have appropriate confidence in the code, but always hate & fear 
'just one more last minute change'.  

Right now I just need to run scalafmt on my code and resolve some conflicts.

**Comment by kitbellew** [comment]
> If it does not cause unnecessary work at your end, I would be happier getting on the
> `2.5.1 ` train. I have appropriate confidence in the code, but always hate & fear
> 'just one more last minute change'.
> 
> Right now I just need to run scalafmt on my code and resolve some conflicts.

Thank you for your work, patience with my (hopefully minor) nitpicking, and let's target 2.5.1.

**Comment by LeeTibbert** [comment]
@kitbellew  I have appreciated your timely & patient review.  Regular expressions are
supposed to save time & complexity but I find they really suck up time.  This PR has 
benefited from the review.  I am sorry that it took so many cycles.

**Comment by LeeTibbert** [comment]
Superseded by PR #1936, its higher & purer form.
